### PR TITLE
enable MKL-DNN/oneDNN usage for aarch64 CPUs in TensorFlow 2.5+

### PR DESCRIPTION
**Pull request by Flamefire**
(created using `eb --new-pr`)

### CODE REVIEW

File: easybuild/easyblocks/t/tensorflow.py [hunk 1]
```diff
@@ -247,6 +246,14 @@ def __init__(self, *args, **kwargs):
 
         self.system_libs_info = None
 
+        if self.cfg['with_mkl_dnn'] is None:
+            if get_cpu_architecture() == X86_64:
+                # Supported on x86 since forever
+                self.cfg['with_mkl_dnn'] = True
+            elif LooseVersion(self.version) >= LooseVersion('2.5.0') and get_cpu_architecture() == AARCH64:
+                # ARM support added in 2.5
+                self.cfg['with_mkl_dnn'] = True
```
Comment (boegel) [inline easybuild/easyblocks/t/tensorflow.py:255]: @Flamefire Let's add some logging here, see https://github.com/Flamefire/easybuild-easyblocks/pull/7

**Review by boegel — CHANGES_REQUESTED** [review]

File: easybuild/easyblocks/t/tensorflow.py [hunk 2]
```diff
@@ -247,6 +246,24 @@ def __init__(self, *args, **kwargs):
...
+            if cpu_arch == X86_64:
+                # Supported on x86 since forever
+                self.cfg['with_mkl_dnn'] = True
+                self.log.info("Auto-enabled use of MKL-DNN on %s CPU architecture", cpu_arch)
+            elif cpu_arch == AARCH64:
+                # ARM support added in 2.5
+                min_tf_ver = '2.5.0'
+                if LooseVersion(self.version) >= LooseVersion(min_tf_ver):
+                    self.cfg['with_mkl_dnn'] = True
+                    self.log.info("Auto-enabled use of MKL-DNN for TensorFlow >= %s on %s CPU architecture", min_tf_ver, cpu_arch)
```
[the remark this answers is not in the document]
Comment (Flamefire) [reply easybuild/easyblocks/t/tensorflow.py:261]:
```suggestion
                    self.log.info("Auto-enabled use of MKL-DNN for TensorFlow >= %s on %s CPU architecture",
                                      min_tf_ver, cpu_arch)
```

**Comment by boegel** [comment]
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **FAIL (build issue)** _TensorFlow-2.5.0-foss-2020b.eb_ (partial log available at https://gist.github.com/5f40f2979bd293569ab1b0c4849533e2)
 * **FAIL (build issue)** _TensorFlow-2.6.0-foss-2021a.eb_ (partial log available at https://gist.github.com/de42ab10b4ea1001ed6af06cc791c45a)

Build succeeded for 0 out of 2 (2 easyconfigs in total)
select-pika-c6gd-2xlarge-0001 - Linux centos linux 8.3.2011, AArch64, ARM UNKNOWN (graviton2), Python 3.6.8
See https://gist.github.com/5a6ba346212cb80fda239c0f1a73198f for a full test report.

**Comment by Flamefire** [comment]
@boegel 
> gcc: fatal error: Killed signal terminated program cc1plus

Any cgroup stuff maybe?

**Comment by boegel** [comment]
> @boegel
> 
> > gcc: fatal error: Killed signal terminated program cc1plus
> 
> Any cgroup stuff maybe?

Or perhaps lack of memory...

This was on a [c6g.2xlarge](https://aws.amazon.com/ec2/instance-types/c6/) AWS instance (8 cores, 16GB of RAM), which is perhaps a bit on the light side to build TensorFlow from source (using 8 cores)?

I'll try again on a fatter node (or using less cores).

**Comment by boegel** [comment]
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **FAIL (build issue)** _TensorFlow-2.5.0-foss-2020b.eb_ (partial log available at https://gist.github.com/e98e1014edb6d420afbd778addd1f6ac)
 * **FAIL (build issue)** _TensorFlow-2.6.0-foss-2021a.eb_ (partial log available at https://gist.github.com/da8ea294f219964cdf480e02b542f845)

Build succeeded for 0 out of 2 (2 easyconfigs in total)
select-pika-c6gd-4xlarge-0001 - Linux centos linux 8.3.2011, AArch64, ARM UNKNOWN (graviton2), Python 3.6.8
See https://gist.github.com/1b7f95f9ca3960d7e3d2a2de5403ce0a for a full test report.

**Comment by Flamefire** [comment]
@boegel Bug in TF with excluding based on arch and a config flag leading to ambigious, yet same select cases: https://github.com/tensorflow/tensorflow/blob/72fd2bfa42a8ad909baf8d2b7b674563d256514d/tensorflow/core/kernels/BUILD#L3028-L3034 (arm_any and no_mkldnn_contraction_kernel triggers)
Can you open a TF bug?

**Comment by boegel** [comment]
@Flamefire It's not clear to me what's going wrong here, so I'm not sure what to put in the issue. Also, I have very little bandwidth to follow up on it...

Does it basically mean that the `--config=mkl_aarch64` doesn't work at all, or is it more subtle than that?

**Comment by Flamefire** [comment]
Did that: https://github.com/tensorflow/tensorflow/issues/52027

More or less: yes. However for ARM we might want to use `--define build_with_mkl_aarch64=true` anyway, see https://github.com/easybuilders/easybuild-easyblocks/issues/2577#issuecomment-919914929
Or None at all, i.e. change `with_mkl_dnn` to `False`, especially as the name is likely misleading...

**Comment by Flamefire** [comment]
Closing as this isn't actually enabling mklDNN/oneDNN. See https://github.com/easybuilders/easybuild-easyblocks/issues/2577 and https://groups.google.com/a/tensorflow.org/g/build/c/RZhgZst-fgQ

**Pull request closed without merging**